### PR TITLE
feat: Adding dp_bitmasking example in dynamic_programming.

### PR DESCRIPTION
**Pull request by faizan2700**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1475/checks?check_run_id=2309561860) warnings.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
The [checks](https://github.com/TheAlgorithms/C-Plus-Plus/pull/1475/checks?check_run_id=2309670413) are still failing. Please fix them.

**Comment by faizan2700** [comment]
Suggestions for improvement are appreciated. Thank you for reviewing.

File: dynamic_programming/dp_bitmasking.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        for(uint8_t j1 = 0; j1 < total_unselected; j1++){
+            for(uint8_t j2 = j1+1; j2 < total_unselected; j2++){
+                uint8_t x = unselected_elements[j1], y = unselected_elements[j2];
+                uint32_t new_mask = i | (1<<x) | (1<<y);
+                dp[new_mask] = std::max(dp[new_mask], dp[i] + (total_pairs + 1) * G[x][y]);
+            }
+        }
+    }
+
+    //Max score is in the mask 2**(1<<2*N) - 1, which shows all elements are added.
```
Comment (Panquesito7) [inline dynamic_programming/dp_bitmasking.cpp:88]:
```suggestion
    // Max score is in the mask 2**(1<<2*N) - 1, which shows all elements are added.
```

File: dynamic_programming/dp_bitmasking.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    return dp[M-1];
+
+}
+
+} //namespace dp_bitmask
+} //namespace dynamic_programming
```
Comment (Panquesito7) [inline dynamic_programming/dp_bitmasking.cpp:94]:
```suggestion
}  // namespace dp_bitmask
}  // namespace dynamic_programming
```

File: dynamic_programming/dp_bitmasking.cpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main(){
+    //Testing
+    test();
```
Comment (Panquesito7) [inline dynamic_programming/dp_bitmasking.cpp:126]:
```suggestion
    test();  // run self-test implementations
```

File: dynamic_programming/dp_bitmasking.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+} //namespace dynamic_programming
+
+
+/**
+ * @brief Function to test above algorithm
+ * @returns void
+ */
+static void test(){
+
+    // Test 1
```
Comment (Panquesito7) [inline dynamic_programming/dp_bitmasking.cpp:103]:
```suggestion
    // 1st test
```

File: dynamic_programming/dp_bitmasking.cpp [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    const uint8_t N1 = 5;
+    std::array<uint32_t, 2*N1> A = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+    uint64_t ans = 0;
+    ans = dynamic_programming::dp_bitmask::find_max_score(A, N1);
+    assert(ans == 55);
+
+    // Test 2
```
Comment (Panquesito7) [inline dynamic_programming/dp_bitmasking.cpp:110]:
```suggestion
    // 2nd test
```

File: dynamic_programming/dp_bitmasking.cpp [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+} //namespace dp_bitmask
+} //namespace dynamic_programming
+
+
+/**
+ * @brief Function to test above algorithm
+ * @returns void
+ */
+static void test(){
+
```
Comment (Panquesito7) [inline dynamic_programming/dp_bitmasking.cpp:102]:
```suggestion
```

File: dynamic_programming/dp_bitmasking.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    return dp[M-1];
+
+}
+
+} //namespace dp_bitmask
+} //namespace dynamic_programming
+
+
+/**
+ * @brief Function to test above algorithm
```
Comment (Panquesito7) [inline dynamic_programming/dp_bitmasking.cpp:98]:
```suggestion
 * @brief Self-test implementations
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work!
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/faizan2700/C-Plus-Plus/actions

**Comment by faizan2700** [comment]
DP - Bitmasking example

Problem : Given an Integer N, and an Array of Integer A of size 2*N. 
Select N pairs from this array such that each element is present in 
exactly one pair, And the following score is maximized.

score = summation(gcd(x_i, y_i) * i) for i 1 to n.

Solution :
See that score depends upon the selection of pairs and order of pairs.
consider the subset of the array which contains 2*n elements and n pairs.
this subset is denoted by binary number where j-th bit is 1 if that element 
is included and 0 if that element is excluded (0101 -> first and third included, 
second and fourth excluded). using this we can find the maximum value we can
get for selection of n+1 pairs, if n-pairs are selected from this pair and one pair 
is made from unselected elements. of course the subset which contains n+1 pairs, 
will be updated by all different subsets of n-pairs and hence will contain the optimal 
value. we can start from mask 0 which shows no pair is selected and keep updating like
this till the mask which shows all elements are selected 1111...(2*N times) that is 
2^(2*N) - 1, this mask will contain the answer. so finally we can write dp transition as


dp[subset of n+1 pairs] = max( dp[subset of size n-pairs] and adding gcd(new_pair from 
unselected elements) )

**Comment by faizan2700** [comment]
Please tell me if this explanation is good. I will add this in bottom of the code file, if allowed.

**Pull request closed without merging**